File: lib/data/model/article.dart
import 'package:freezed_annotation/freezed_annotation.dart';
import 'package:quiz_shower/data/model/firestore_datetime_convert.dart';

part 'article.freezed.dart';
part 'article.g.dart';

@freezed
class Article with _$Article {
  const Article._();

  const factory Article({
    required String id,
    required String createdBy,
    required String title,
    required String description,
    required bool isBookmarked,
    required ArticleType type,
    required String? url,
    required String content,
    @Default(false) bool isPublic,
    @FirestoreDateTimeConverter() required DateTime createdAt,
  }) = _Article;

  factory Article.fromJson(Map<String, dynamic> json) => _$ArticleFromJson(json);

  static Article mock() {
    return Article(
      id: '1',
      createdBy: 'taro',
      createdAt: DateTime.now(),
      title: '透明UIを作る',
      description: 'Flutterで透明UIを作る方法を紹介します。',
      isBookmarked: false,
      type: ArticleType.url,
      url: 'https://gajeroll.com/20230606-wwdc23/',
      content: 'Flutterで透明UIを作る方法を紹介します。',
    );
  }
}

enum ArticleType {
  url,
  manual,
}

File: functions/src/types/article.ts
import * as admin from "firebase-admin";
import { Timestamp } from "firebase-admin/firestore";

export type Article = {
  id: string;
  createdBy: string;
  createdAt: Timestamp;
  title: string;
  description: string;
  isBookmarked: boolean;
  type: "url" | "manual";
  content: string;
  isPublished: boolean;
};

export const articleConverter = {
  toFirestore(article: Article): admin.firestore.DocumentData {
    return {
      id: article.id,
      createdBy: article.createdBy,
      createdAt: article.createdAt,
      title: article.title,
      description: article.description,
      isBookmarked: article.isBookmarked,
      type: article.type,
      content: article.content,
      isPublished: article.isPublished,
    };
  },
  fromFirestore(snapshot: admin.firestore.QueryDocumentSnapshot): Article {
    const data = snapshot.data();

    return {
      id: data.id,
      createdBy: data.createdBy,
      createdAt: data.createdAt,
      title: data.title,
      description: data.description,
      isBookmarked: data.isBookmarked,
      type: data.type,
      content: data.content,
      isPublished: data.isPublished,
    };
  },
};
